<!doctype html>
<html lang='en'>
  <head>
    <meta charset='utf-8'>

    <title>Automatic Revision of Ecological Theories</title>

    <meta name='description' content='Thesis defense'>
    <meta name='author' content='Philippe Desjardins-Proulx'>

    <meta name='apple-mobile-web-app-capable' content='yes' />
    <meta name='apple-mobile-web-app-status-bar-style' content='black-translucent' />

    <meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no'>

    <link rel='stylesheet' href='css/reveal.min.css'>
    <link rel='stylesheet' href='css/theme/thesis.css' id='theme'>
    <link rel='stylesheet' href='lib/css/zenburn.css'>

    <script>
      if (window.location.search.match(/print-pdf/gi)) {
        var link = document.createElement('link');
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = 'css/print/pdf.css';
        document.getElementsByTagName('head')[0].appendChild(link);
      }
    </script>

    <!--[if lt IE 9]>
    <script src='lib/js/html5shiv.js'></script>
    <![endif]-->
  </head>

  <body>

    <div class='reveal'>
      <!-- Any section element inside of this container is displayed as a slide -->
      <div class='slides'>
        <section>
          <section>
            <h1>Automatic Revision<br/> of Ecological Theories</h1>
            <p>by <a href='http://phdp.github.io/'>Philippe Desjardins-Proulx</a></small></p>
          </section>
        </section>
        <section>
          <section>
            <h1>Motivation</h1>
          </section>
          <section>
            <img src='img/amnat12.png' alt='amnat12'/>
          </section>
          <section>
            <img src='img/eco-evo-paper.png' alt='ecoevo'/>
          </section>
          <section>
            <h2>Species Richness/Speciation Relationship</h2>
            <img width='60%' height='60%' src='img/complex-1.svg' alt='res'/>
          </section>
          <section>
            <img src='img/hahn08.png' alt='hahn08'/>
          </section>
          <section>
            <img src='img/gillespie1997.png' alt='gil04'/>
            <p><i>Reference: JH Gillespie. Population Genetics: A Concise Guide, 1st edition. Johns Hopkins University Press, 1997.</i></p>
          </section>
          <section>
            <h2>Why Biodiversity is so Complicated?</h2>
            <p>The effect of mutations vary greatly accross a single gene.</p>
            <p>Selection accross the genome is influenced by linkage and recombination.</p>
            <p>Selection varies in time/space as the biotic/abiotic environment changes.</p>
            <p>Speciation has a complex spatial structure.</p>
          </section>
          <section>
            <h2>What if there is no<br/>single theory of biodiversity?</h2>
          </section>
          <section>
            <h2>What if the problem is to allow representation of a set of ecological theories, and tools to revise them.</h2>
          </section>
          <section>
            <h2>Machine Learning</h2>
            <img src='img/xkcd_machine_learning.png' alt='xkcd_ml'/>
            <p>Two main types: supervised learning, unsupervised learning.</p>
            <p><i>Reference: KP Murphy. Machine Learning: a Probabilistic Perspective. MIT Press, 2012.</i></p>
          </section>
          <section>
            <h2>Supervised Learning</h2>
            <img>
            <p>Given data points \(\mathcal{D} = \{(\mathbf{x}_i, y_i)\}_{i = 0}^{|\mathcal{D}| - 1}\), learn to predict \(y\) from \(\mathbf{x}\).</p>
            <p>Tasks: <i>classification, regression.</i></p>
          </section>
          <section>
            <h2>Unsupervised Learning</h2>
            <img>
            <p>Given data points \(\mathcal{D} = \{(\mathbf{x}_i)\}_{i = 0}^{|\mathcal{D}| - 1}\), learn some 'hidden structure'.</p>
            <p>Tasks: <i>matrix imputation, structure learning, clustering.</i></p>
          </section>
          <section>
            <h2>A knowledge base for ecological knowledge</h2>
            <p><b>Domains</b>: species \(s, s_i, s_j \in S\), locations \(l \in L\), time \(t \in \mathbb{R}\).</p>
            <p><b>Symbols</b>: universal quantifier (<i>forall</i>) \(\forall\), negation \(\neg\).</p>
            <br/>
            \[\forall s_i, s_j: PhyloDistance(s_i, s_j) > 0, \infty.\]
            \[\forall s_i, s_j: \dot{x} = \alpha s_i - \beta s_is_j \mbox{ and } \dot{s_j} = \delta s_is_j - \gamma s_j, 0.01.\]
            <p>...</p>
            \[\forall s, l, t: SmallN(s, l) \Rightarrow N(s,l,t+1) = G(s) \times N(s,l,t), 0.14.\]
            \[\forall s_i, s_j, l: KNN_5(s_i, s_j) > 0.5 \Rightarrow PreyOnAt(p, q, l), 2.5.\]
          </section>
          <section>
            <h2>Why ecological interactions?</h2>
            <p>First reason: a key reason why selection is difficult to handle
            with traditional theories is because of the complexity of species
            interactions.</p>
            <br/>
            <p>Second reason: my lab is studying interactions. I had access to
            both good data-sets and people who understood them well.</p>
          </section>
          <section>
            <h2>Ecological interactions as a machine learning problem</h2>
            <p>Predict whether \(a\) interacts with \(b\) given traits for both species.</p>
            <p>Fill holes in an interaction matrix (Matrix imputation).</p>
            <p>Discover or revise logic and mathematical rules given data.</p>
          </section>
          <section>
            <h2>Roadmap</h2>
            <img src='img/roadmap.svg.png' alt='roadmap'/>
          </section>
        </section>
        <section>
          <section>
            <h1>Classical Machine<br/>Learning</h1>
          </section>
          <section>
            <h2>Data</h2>
            <ul width='100%'>
              <li>Assembled by Isabelle Daigle, mostly from Digel et al. (2014).</li>
              <li>48 soil food webs.</li>
              <li>881 species.</li>
              <li>34 193 unique interactions.</li>
              <li>215 418 absence of interactions.</li>
              <li>3 real valued-traits: \(body mass\), \(Ph0\), \(Ph1\).</li>
              <li>25 binary traits (e.g. AboveGround, Detritivore, Bacteria, Jumps, LongLegs, UsePoison).</li>
            </ul>
          </section>
          <section>
            <img src='img/ecological-netflix-paper.png' alt='paper'/>
          </section>
          <section>
            <img src='img/pearson.svg' alt='pearson'/>
          </section>
          <section>
            <h2>Similarity measure</h2>
            \[tanimoto(\mathbf{x}, \mathbf{y}) = \frac{\left\vert\mathbf{x} \cap \mathbf{y}\right\vert}{\left\vert\mathbf{x} \cup \mathbf{y}\right\vert}.\]

            <br/>
            <p>For \(\mathbf{x} = \{A, B, C\}\) and \(\mathbf{y} = \{A, C, D, E, F\}\):</p>
            <br/>

\begin{align}
  tanimoto(\mathbf{x}, \mathbf{y}) &= \frac{\left\vert\{A, B, C\} \cap \{A, C, D, E, F\}\right\vert}{\left\vert\{A, B, C\} \cup \{A, C, D, E, F\}\right\vert}\\
                                   &= \frac{\left\vert\{A, C\}\right\vert}{\left\vert\{A, B, C, D, E, F\}\right\vert} = \frac{2}{6}.
\end{align}


          </section>
          <section>
            <h2>True Skill Statistic</h2>
            \[TSS = \frac{(tp \times tn) - (fp \times fn)}{(tp + fn)(fp + tn)}.\]

<br/><br/>

            <p><i>tp, tn, fp, fn</i> are the true positives, true negatives, false positives,
            and false negatives, respectively.</p>

            <p>The <i>TSS</i> ranges from -1 to 1.</p>

            <p>CONFUSION MATRIX.</p>
          </section>
          <section>
            <h2>Results</h2>
            <p>Supervised learning Random forests predict correctly 99.55% of
            the non-interactions and 96.81% of the interactions, for a
            <i>TSS</i> of 0.96.</p>

            <br/>

            <p>Removing the binary traits has little effect on the model. With
            only \(body mass\), \(Ph0\), \(Ph1\), the <i>TSS</i> of the random
            forest is 0.94.</p>
          </section>
          <section>
            <img src='img/knn-paper.png' alt='knn'/>
            <p>Larger!!!!</p>
          </section>
          <section>
            <h2>I Bartomeus' pollinator-plant data-set</h2>
            <ul width='100%'>
              <li>Plant-pollinator interactions were collected across 16 sites in
              the southwest of the iberian peninsula (Huelva and Seville).</li>
              <li>65 pollinators.</li>
              <li>277 plants.</li>
              <li>739 pollinator-plant interactions.</li>
              <li>We randomly pick 739 pollinator-plant pairs that are not
              interacting to build a data-set of 1478 entries, half of
              interactions, half of non-interactions.</li>
            </ul>
          </section>
          <section>
            <h2>Accuracy</h2>
            <img src='img/ht_acc.svg' width='80%' alt='acc'/>
          </section>
          <section>
            <h2>TSS</h2>
            <img src='img/ht_tss.png' width='80%' alt='acc'/>
          </section>
          <section>
            <h2>Classical machine learning &amp; Interactions</h2>
            <ul>
              <li>We confirm the old result that body mass is key to how interactions are structured (Williams &amp; Martinez 2000).</li>
              <li>Phylogeny an important feature.</li>
              <li>Interactions are fairly easy to predict (TSS really close to 1).</li>
              <li>Binary traits contain little useful information.</li>
            </ul>
          </section>
        </section>
        <section>
          <section>
            <h1>Markov Logic</h1>
          </section>
          <section>
            \[\forall x, y, z: Friends(x, y) \mbox{ and } Friends(y, z) \Rightarrow Friends(x, z), 0.7.\]
            \[\forall x: Smoking(x) \Rightarrow Cancer(x), 1.5.\]
            \[\forall x, y: Friends(x, y) \mbox{ and } Smoking(x) \Rightarrow Smoking(y), 1.1.\]

            <br/><br/>

            <p>Can answer queries of the form:</p>

            <br/>

            \[P(Cancer(Bob) | Smoking(Anna) \mbox{ and } Friends(Anna, Bob)).\]
            \[P(Friends(x, y) \Leftrightarrow Friends(y, x) | ...).\]

          </section>
          <section>
            <h2>Salix data-set</h2>
            <img src='img/salix.png' alt='salix'/>
            \[Parasites \rightarrow Galler \rightarrow Salix\]
          </section>
          <section>
            <h2>Learning weights (but not really rules)</h2>
<p>\(\forall x, y, z: PreyOnAt(x,y,z) \Rightarrow PreyOn(x,y), 5.20.\)</p>
<p>\(\forall x, y: HighCooccurrence(x,y) \Rightarrow PreyOn(x,y), 4.22.\)</p>
<p>\(\forall x, y: IsGaller(x) \mbox{ and } PreyOn(x,y) \Rightarrow IsSalix(y), 4.15.\)</p>
<p>\(\forall x, y: IsParasitoid(y) \mbox{ and } PreyOn(y,x) \Rightarrow IsGaller(x), 3.49.\)</p>
<p>\(\forall x, y: PreyOn(x,y) \Rightarrow HighCooccurrence(x,y), 1.57.\)</p>
<p>\(\forall x, y: PreyOn(x,y) \Rightarrow PreyOn(x,x), 1.52.\)</p>
<p>\(\forall x, y: HigherPhyloValue(x,y) \Rightarrow PreyOn(x,y), 1.13.\)</p>
<p>\(\forall x, y: HigherPhyloValue(x,y) \Rightarrow HighCooccurrence(x,y), -0.59.\)</p>
<p style='color:#BE1F1F'>\(\forall x, y: PreyOn(x,x) \Rightarrow PreyOn(x,y), 0.45.\)</p>
<p style='color:#BE1F1F'>\(\forall x, y: \neg IsSalix(x) \mbox{ or } \neg PreyOn(x,y), 0.02.\)</p>
<p>\(\forall x, y, z: CloselyRelated(x,y) \mbox{ and } PreyOn(x,z) \Rightarrow PreyOn(y,z), 0.00.\)</p>
          </section>
          <section>
            <h2>What went wrong?</h2>
            <p>Lessons from classical methods: real-valued variables are important.</p>
            <p>Importance of nuances. PreyOn, ... Lessons from last project.</p>
          </section>
        </section>
        <section>
          <section>
            <h1>Fuzzy Logic</h1>
          </section>
          <section>
            <h2>I Bartomeus' pollinator-plant data-set (again)</h2>
            <ul width='100%'>
              <li>Plant-pollinator interactions were collected across 16 sites in the southwest of the iberian peninsula (Huelva and Seville).</li>
              <li>65 pollinators.</li>
              <li>277 plants.</li>
              <li>739 pollinator-plant interactions.</li>
              <li>We randomly pick 739 pollinator-plant pairs that are not
              interacting to build a data-set of 1478 entries, half of
              interactions, half of non-interactions.</li>
            </ul>
          </section>

          <section>
            <h2>Structure</h2>
            \[\mathbf{If} \mbox{ } antecedants \mbox{ } \mathbf{then} \mbox { } consequents.\]
            \[\mathbf{If} \mbox{ } antecedants \mbox{ } \mathbf{then} \mbox { } consequents.\]
            \[\mathbf{If} \mbox{ } antecedants \mbox{ } \mathbf{then} \mbox { } consequents.\]
            \[...\]

            <br/>
            <br/>

            \[\mathbf{If} \mbox{ } A_0 \mbox{ and } A_1 \mbox{ and } A_2 \mbox{ and } ... \mbox{ } \mathbf{then} \mbox { } C_0.\]
          </section>
          <section>
            <h2>Linguistic variables</h2>
            <img width='70%' height='60%' src='img/fsets.svg'/>
            <p>If the flower width is 6.5mm, then <b>"Flower width is High"</b> has a truth value of 0.75, <b>"Flower width is Very High"</b> has a truth alue of 0.25.</p>
          </section>
          <section>
            <h2>Initial model for I Bartomeus' dataset</h2>
            <ul style='width: 100%'>
              <li style='color:#277F2A'><b>If</b> KNN is high <b>then</b> \(I^+\).</li>
              <li style='color:#7F2D27'><b>If</b> KNN is low <b>then</b> \(I^-\).</li>
            </ul>
            </br>
            <p>\(I^+\) denotes an interaction, \(I^+\) a non-interaction.</p>
          </section>
          <section>
            <h2>Learning rules</h2>
            <ol start='0'>
              <li>Start with the two KNN rules.</li>
              <li>Randomly generate a rule, add it to the knowledge base.</li>
              <li>If the new knowledge base beats the old by \(\alpha\) TSS on the training set, keep the rule.</li>
              <li>Repeat #1-2 \(\phi\) times.</li>
            </ol>
            <br/>
            <br/>
            <p>Starts with high acceptance threshold \(\alpha\), lower it with time.</p>
            <br/>
            <p>C++14 code: <a href='https://github.com/PhDP/Manticore'>https://github.com/PhDP/Manticore</a></p>
          </section>
          <section>
            <ul style='width: 100%; font-size: 80%'>
              <li style='color:#277F2A'><b>If</b> Plant nectar tube depth is average <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Pollinator body size is small <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> KNN is high <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Pollinator is specialist <b>and</b> Plant nectar tube dim is small <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Plant flower width is average <b>then</b> \(I^+\).</li>
              <li style='color:#277F2A'><b>If</b> Plant flower width is small <b>and</b> Pollinator tongue length is small <b>then</b> \(I^+\).</li>
              <li style='color:#7F2D27'><b>If</b> KNN is low <b>then</b> \(I^-\).</li>
            </ul>
          </section>
          <section>
            <h2>Results on pollinator-plant data-set</h2>
            <table style='width:100%; font-size:80%; text-align:center;'>
              <tr>
                <th>Model</th>
                <th>TSS (10% testing data)</th>
              </tr>
              <tr>
                <td>Decision trees</td>
                <td>0.723</td>
              </tr>
              <tr>
                <td>Random forests</td>
                <td>0.795</td>
              </tr>
              <tr>
                <td>Fuzzy (initial model with two KNN rules)</td>
                <td>0.564</td>
              </tr>
              <tr>
                <td>Fuzzy (after evolution)</td>
                <td>0.885</td>
              </tr>
            </table>
            <p>The <a href='http://scikit-learn.org/stable/index.html'>scikit-learn</a> Python package was used for decision trees and random forests.</p>
          </section>
        </section>
        <section>
          <section>
            <h1>Conclusions</h1>
          </section>
          <section>
            <ul width='100%'>
              <li>KNN is an effective, widely applicable, and simple approach.</li>
              <li>KNN with threshold works well with TSS scores.</li>
              <li>Supervised learning algorithms (esp. random forests) can predict interactions with few real-valued traits.</li>
              <li>Fuzzy knowledge bases are both clear and accurate.</li>
              <li>Even a silly revision algorithm yields very good results for fuzzy logic.</li>
              <li>Knowledge representation in ecology is an interesting topic in itself.</li>
              <li>Markov logic has issues (computational) but modern variants could be promising.</li>
            </ul>
          </section>
        </section>
        <section>
          <section>
            <h1>Questions?</h1>
          </section>
        </section>

      </div>
    </div>

    <script src='lib/js/head.min.js'></script>
    <script src='js/reveal.min.js'></script>
    <script>
      // Full list of configuration options available here:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        width: '70%',
        height: '100%',
        margin: 0,
        minScale: 1,
        maxScale: 1,
        controls: true,
        progress: true,
        history: true,
        center: true,

        theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
        transition: Reveal.getQueryHash().transition || 'linear', // default/cube/page/concave/zoom/linear/fade/none

        // Parallax scrolling
        // parallaxBackgroundImage: 'https://s3.amazonaws.com/hakim-static/reveal-js/reveal-parallax-1.jpg',
        // parallaxBackgroundSize: '2100px 900px',

        // Optional libraries used to extend on reveal.js
        dependencies: [
          { src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
          { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } },
          { src: 'plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
          { src: 'plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } }
        ]
      });
    </script>
    <script type='text/javascript' src='../../js/bower/MathJax/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
